### PR TITLE
SCons: Add `tests` option to enable or disable unit tests

### LINKED ISSUE #107: Add TESTS_ENABLED option to run tests in `release` mode and ability to opt-out from compiling tests
Might help to resolve GDScript module dependencies as described in #117.
Namely: godotengine/godot#31011.

**Describe the project you are working on:**
A project in which running engine-specific tests is not crucial for development.

`[Ad]` For project-specific testing I highly recommend: [`GUT`](https://github.com/bitwes/Gut) [`WAT`](https://github.com/CodeDarigan/WAT/).

**Describe how this feature / enhancement will help your project:**
First, see the PR discussion at godotengine/godot#32365 which ended up as mainly consistency fix. In there @santouits expressed concern that tests should be treated separately, and I think that makes sense for variety of reasons:

1. We should actually care about tests.

2. Ability to run tests in `release` mode, which is currently not possible as tests are only being compiled with `tools=yes`. This can help to save bugs and regressions to occur, and especially relevant for GDScript module where conditional compilation has lead to numerous hidden bugs in `release` mode while they run perfectly fine in `debug` or `release_debug`:

* godotengine/godot#25408
* godotengine/godot#25334
* godotengine/godot#25261
* godotengine/godot#25124
* godotengine/godot#24464
* godotengine/godot#28680

3. Have an option to opt-out from compiling tests to speed up compiling Godot source. It doesn't currently have much impact but if we write more tests, that could slow down build time to some extent. Users don't really care about tests if they're not developing Godot itself.

4. Ability to disable tests compilation could also improve some CI builds with limited resources as the amount of PRs being created each day.  

    Alternitavely, we could just port C++ tests to GDScript and move them into [`godot-tests`](https://github.com/godotengine/godot-tests) repository being self-contained projects and tested in CI environment. In fact this is already done for `server` build with minimal project being exported.

    On the other hand, developers would find it more cumbersome to clone `godot-tests` repository just to run tests, and there's also an issue of maintaining tests independently of main repository: https://github.com/godotengine/godot/issues/2641#issuecomment-372026325.

**Show a mock up screenshots/video or a flow diagram explaining how your proposal will work:**
See below.

**Describe implementation detail for your proposal (in code), if possible:**
See preliminary changes at godotengine/godot#32387.

I suggest adding `scons tests=yes` command line option which would work independently from `target`/`tools` in root `Sconstruct`, so it would be used like so:
```
scons target=release tests=yes bits=64 -j8
```
Tests could be enabled by default if target is `debug`, for instance.

There should be made necessary changes to `main.cpp` to replace `DEBUG_ENABLED` or `TOOLS_ENABLED` conditional compilation to have `TESTS_ENABLED` on for test-related code. The mentioned PR demonstrates how the absence of dedicated macro has lead to confusion and therefore link errors.

**If this enhancement will not be used often, can it be worked around with a few lines of script?:**
This is related to the buildsystem, can't be worked around.

**Is there a reason why this should be core and not an add-on in the asset library?:**
This is related to the buildsystem, and it's part of the core already.

### PR DESCRIPTION
**Pull request by Xrayez**
Closes godotengine/godot-proposals#107.
Helps to workaround #40666.
Remake of #32387.

### CODE REVIEW

File: SConstruct [hunk 1]
```diff
@@ -115,6 +115,7 @@ opts.Add(EnumVariable("target", "Compilation target", "debug", ("debug", "releas
 opts.Add(EnumVariable("optimize", "Optimization type", "speed", ("speed", "size")))
 
 opts.Add(BoolVariable("tools", "Build the tools (a.k.a. the Godot editor)", True))
+opts.Add(BoolVariable("tests", "Build the unit tests", True))
```
Comment (akien-mga) [inline SConstruct:118]: I think `tests` should be only enabled on `tools` builds, I don't think they would work on export templates anyway as is.

And maybe the default value should be `False` for end users who build the engine, and make `dev=yes` enable it.

**Comment by RevoluPowered** [comment]
Can we edit the main/tests/scsub so that if tests=no nothing is done in that directory compilation wise?

**Comment by Xrayez** [comment]
Yeah this is already the case, but the check is within `main/SCsub` which calls into `tests/SCsub`, building with `tests=no` would prevent going into that dir.
```python
if env["tests"]:
    SConscript("tests/SCsub")

```

**Comment by Xrayez** [comment]
Added `tests=yes` to CI configs because tests are disabled by default now, as suggested.

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!

**Pull request merged**